keycloak_api_client/api_client.py

```diff
@@ -189,24 +193,38 @@ def _get_user_endpoint_schema_data(
 
         return data
 
+    def _get_keycloak_user(
+        self,
+        user_endpoint_data: dict
+    ) -> Union[ReadKeycloakUser, object]:
+        keycloak_user = read_keycloak_user_factory(
+            user_endpoint_data=user_endpoint_data
+        )
+
+        if self.keycloak_user_factory_wrapper:
+            return self.keycloak_user_factory_wrapper(keycloak_user)
```

Comment:
Why do we have to incorporate such things in the API client? Isn't the caller responsibility to convert `ReadKeycloakUser`?

Reply:
@wsz this simplifies using library. If you just need standard `ReadKeycloakUser` then you're fine without it, but if you have your own customized user attributes in Keycloak you may want e.g. receive `YourExtendedReadKeycloakUser` what has additional attribute `.phone`. Without optional wrapper you would need to or extend library api client or create additional layer that would be called for all executions of get, search etc. Imo that is more developer friendly to just create wrapper on your side and supply that as a config param without creating layer in between.

Reply:
Ok, I see. I personally prefer low coupling more than usability. Just found there's `raw_data` property in `ReadKeycloakUser`. We won't be able to change it, even the schema of a dict inside, if we don't want to break consumer's code. That's something worth a comment or even [ADR](https://masterplan.slab.com/topics/decisions-romjzc9v).

Reply:
@wsz I'm not sure I understand what's your idea around `raw_data`? It is a dump from user endpoint answer in case of need of any information that was not serialized to `ReadKeycloakUser` object (e.g. token scopes).

Reply:
Let's suppose somebody would like to use `keycloak_user_factory_wrapper` to receive `MyReadKeycloakUser` instead of `ReadKeycloakUser`. Following the use case provided by you (`.phone`) somebody has to use `raw_data` which implies this property should always exist and should contain a key, for example, `phone_raw_unformatted`. We can't change that if we don't want to break consumer's code. The schema of `raw_data` becomes a part of the public interface. And it's not obvious enough, I think.

Reply:
@wsz do you have some solution for that in mind? Raw data is pure endpoint response which is some JSON. It should be consistent, but it depends on Keycloak rest api schema which may change anyways.

Reply:
Just add a warning comment above `raw_data` declaration saying its schema corresponds with [Keycloak UserRepresentation](https://www.keycloak.org/docs-api/8.0/rest-api/index.html#_userrepresentation), and it's required by `keycloak_user_factory_wrapper`.
BTW you should loose some typings:
`keycloak_user_factory_wrapper: Callable[[ReadKeycloakUser], Any] = None` `Optional` is not needed at all, `Any` is more appropriate because why not be more general (what if somebody would like to serialize the output to `str`? typings won't stop him and seems it would work)
However, this implies methods like `search_users(self, query: str)` should return `Any` or `list` because we don't know what we will get
or is there any mandatory interface of the return object? :smirk: With small delegation comes great uncertainty :smile:
All these comments doesn't matter if the lib is used and will be used only by Masterplan.